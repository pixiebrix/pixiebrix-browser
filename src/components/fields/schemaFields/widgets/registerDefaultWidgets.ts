/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import UrlMatchPatternWidget from "./UrlMatchPatternWidget";
import UrlPatternWidget from "@/pageEditor/components/UrlPatternWidget";
import ArrayWidget from "./ArrayWidget";
import BooleanWidget from "./BooleanWidget";
import IntegerWidget from "./IntegerWidget";
import NumberWidget from "./NumberWidget";
import ObjectWidget from "./ObjectWidget";
import OmitFieldWidget from "./OmitFieldWidget";
import SchemaSelectWidget from "./SchemaSelectWidget";
import TemplateToggleWidget from "./TemplateToggleWidget";
import TextWidget from "./TextWidget";
import UnsupportedWidget from "./UnsupportedWidget";
import widgetsRegistry from "./widgetsRegistry";
import WorkshopMessageWidget from "./WorkshopMessageWidget";
import SelectorMatchWidget from "@/pageEditor/components/SelectorMatchWidget";
import CssClassWidget from "./CssClassWidget";
import DatabaseWidget from "./DatabaseWidget";
import PasswordWidget from "./PasswordWidget";
import IconWidget from "@/components/fields/IconWidget";

const defaultWidgets = {
  ArrayWidget,
  BooleanWidget,
  IntegerWidget,
  NumberWidget,
  ObjectWidget,
  OmitFieldWidget,
  SchemaSelectWidget,
  TemplateToggleWidget,
  TextWidget,
  UnsupportedWidget,
  UrlMatchPatternWidget,
  UrlPatternWidget,
  SelectorMatchWidget,
  CssClassWidget,
  WorkshopMessageWidget,
  DatabaseWidget,
  IconWidget,
  PasswordWidget,
} as const;

function registerDefaultWidgets() {
  Object.assign(widgetsRegistry, defaultWidgets);
  for (const [name, widget] of Object.entries(defaultWidgets)) {
    if (!widget) {
      throw new Error(
        `Error registering default widgets. ${name} is undefined. Is there a circular dependency?`
      );
    }
  }
}

export default registerDefaultWidgets;
