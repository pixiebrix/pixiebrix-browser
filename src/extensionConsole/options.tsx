/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/theme/app/app.scss";
import "@/vendors/overrides.scss";
import "@/utils/layout.scss";
import "./options.scss";

import "@/extensionContext";
import "@/development/darkMode";

import { render } from "react-dom";
import React from "react";
import App from "@/extensionConsole/App";
import initGoogle from "@/contrib/google/initGoogle";
import { initToaster } from "@/utils/notify";
import { initTelemetry } from "@/background/messenger/api";
import registerMessenger from "@/extensionConsole/messenger/registration";
import { initPerformanceMonitoring } from "@/telemetry/performance";

function init(): void {
  render(<App />, document.querySelector("#container"));
}

registerMessenger();
void initGoogle();
initToaster();
initTelemetry();
void initPerformanceMonitoring();
init();
