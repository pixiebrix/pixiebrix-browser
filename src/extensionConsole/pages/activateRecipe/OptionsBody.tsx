/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { type RecipeDefinition } from "@/types/recipeTypes";
import genericOptionsFactory from "@/components/fields/schemaFields/genericOptionsFactory";
import FieldRuntimeContext, {
  type RuntimeContext,
} from "@/components/fields/schemaFields/FieldRuntimeContext";
import { OPTIONS_DEFAULT_RUNTIME_API_VERSION } from "@/common";

const OPTIONS_FIELD_RUNTIME_CONTEXT: RuntimeContext = {
  apiVersion: OPTIONS_DEFAULT_RUNTIME_API_VERSION,
  allowExpressions: false,
};

const OptionsBody: React.FunctionComponent<{
  blueprint: Pick<RecipeDefinition, "options">;
}> = ({ blueprint }) => {
  const OptionsGroup = useMemo(
    () =>
      genericOptionsFactory(
        blueprint.options.schema,
        blueprint.options.uiSchema
      ),
    [blueprint.options.schema, blueprint.options.uiSchema]
  );

  return (
    <FieldRuntimeContext.Provider value={OPTIONS_FIELD_RUNTIME_CONTEXT}>
      <OptionsGroup name="optionsArgs" />
    </FieldRuntimeContext.Provider>
  );
};

export default OptionsBody;
