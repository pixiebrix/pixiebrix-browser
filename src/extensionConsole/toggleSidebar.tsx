export const SIDEBAR_ID = "sidebar";
const smallScreenMediaQuery = "(max-width: 991px)";
const sidebarActiveClassName = "active"; // Used to show/hide navbar on small screen
const sidebarIconOnlyClassName = "sidebar-icon-only"; // Used to show/hide navbar on big screen

function getSidebar(): Element {
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- We know it exists
  return document.querySelector(`#${SIDEBAR_ID}`)!;
}

export const toggleSidebar = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.toggle(sidebarActiveClassName);
  } else {
    document.body.classList.toggle(sidebarIconOnlyClassName);
  }
};

export const closeSidebarOnSmallScreen = () => {
  if (window.matchMedia(smallScreenMediaQuery).matches) {
    getSidebar().classList.remove(sidebarActiveClassName);
  }
};
