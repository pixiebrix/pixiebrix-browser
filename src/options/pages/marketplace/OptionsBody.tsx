/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { Card } from "react-bootstrap";
import { RecipeDefinition } from "@/types/definitions";
import genericOptionsFactory from "@/components/fields/schemaFields/genericOptionsFactory";
import FieldRuntimeContext, {
  RuntimeContext,
} from "@/components/fields/schemaFields/FieldRuntimeContext";

// Use "v2" because the service configuration form expects literal values for everything. (I.e., expressions are not
// supports). But we still want to get our SchemaField support for enums, etc.
const OPTIONS_FIELD_RUNTIME_CONTEXT: RuntimeContext = {
  apiVersion: "v2",
};

export interface OptionsBodyProps {
  blueprint: Pick<RecipeDefinition, "options">;
}

const OptionsBody: React.FunctionComponent<OptionsBodyProps> = ({
  blueprint,
}) => {
  const OptionsGroup = useMemo(
    () =>
      genericOptionsFactory(
        blueprint.options.schema,
        blueprint.options.uiSchema
      ),
    [blueprint.options.schema, blueprint.options.uiSchema]
  );
  return (
    <>
      <Card.Body className="px-3 py-3">
        <Card.Title>Personalize Blueprint</Card.Title>
      </Card.Body>
      <Card.Body className="OptionsBody p-3">
        <FieldRuntimeContext.Provider value={OPTIONS_FIELD_RUNTIME_CONTEXT}>
          <OptionsGroup name="optionsArgs" />
        </FieldRuntimeContext.Provider>
      </Card.Body>
    </>
  );
};

export default OptionsBody;
