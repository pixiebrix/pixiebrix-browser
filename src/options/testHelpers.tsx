/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import settingsSlice from "@/store/settingsSlice";
import { blueprintModalsSlice } from "@/options/pages/blueprints/modals/blueprintModalsSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import blueprintsSlice from "@/options/pages/blueprints/blueprintsSlice";
import { recipesSlice } from "@/recipes/recipesSlice";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      settings: settingsSlice.reducer,
      options: extensionsSlice.reducer,
      blueprintModals: blueprintModalsSlice.reducer,
      blueprints: blueprintsSlice.reducer,
      recipes: recipesSlice.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
