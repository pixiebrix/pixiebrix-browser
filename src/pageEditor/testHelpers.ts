/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
