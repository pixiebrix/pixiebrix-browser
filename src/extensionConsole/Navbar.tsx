/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Navbar.module.scss";

import React, { useEffect } from "react";
import { Nav } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBars, faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";
import { Link } from "react-router-dom";
import { getBaseURL } from "@/services/baseService";
import { useAsyncState } from "@/hooks/common";
import {
  addListener as addAuthListener,
  readPartnerAuthData,
  removeListener as removeAuthListener,
} from "@/auth/token";
import { useSelector } from "react-redux";
import { toggleSidebar } from "./toggleSidebar";
import { type SettingsState } from "@/store/settings/settingsTypes";
import cx from "classnames";
import { selectAuth } from "@/auth/authSelectors";
import { type ThemeLogo } from "@/themes/themeUtils";
import useLinkState from "@/auth/useLinkState";
import { DEFAULT_SERVICE_URL } from "@/urlConstants";

function useAdminConsoleUrl(): string {
  // Need to update serviceURL on changes to partner auth data:
  // https://github.com/pixiebrix/pixiebrix-extension/issues/4594
  const [serviceURL, _serviceUrlPending, _serviceUrlError, refreshServiceUrl] =
    useAsyncState<string>(async () => {
      const baseURL = await getBaseURL();
      const partnerAuth = await readPartnerAuthData();
      const url = partnerAuth?.token
        ? new URL("partner-auth", baseURL)
        : new URL(baseURL);
      return url.toString();
    });

  useEffect(() => {
    // Listen for token invalidation
    const handler = async () => {
      void refreshServiceUrl();
    };

    addAuthListener(handler);

    return () => {
      removeAuthListener(handler);
    };
  }, [refreshServiceUrl]);

  return serviceURL ?? DEFAULT_SERVICE_URL;
}

const Navbar: React.FunctionComponent<{ logo: ThemeLogo }> = ({ logo }) => {
  const { email } = useSelector(selectAuth);

  const { hasToken: connected, tokenLoading: connectedPending } =
    useLinkState();

  const serviceURL = useAdminConsoleUrl();

  const mode = useSelector<{ settings: SettingsState }, string>(
    ({ settings }) => settings.mode
  );

  // Use `connectedPending` to optimistically show the toggle
  const showNavbarToggle = mode === "local" || connected || connectedPending;

  return (
    <nav className="navbar default-layout-navbar col-lg-12 col-12 p-0 fixed-top d-flex flex-row">
      <div className={cx(styles.collapsedWrapper, "navbar-brand-wrapper")}>
        <Link className="navbar-brand brand-logo" to="/">
          <img src={logo.regular} alt="PixieBrix logo" />
        </Link>
        <Link className="navbar-brand brand-logo-mini" to="/">
          <img src={logo.small} alt="PixieBrix mini logo" />
        </Link>
        {showNavbarToggle && (
          <button
            className={cx("navbar-toggler", styles.collapsedSidebarToggler)}
            type="button"
            onClick={toggleSidebar}
          >
            <FontAwesomeIcon icon={faBars} />
          </button>
        )}
      </div>
      <div className={cx(styles.expandedWrapper, "navbar-menu-wrapper")}>
        {showNavbarToggle && (
          <button
            className="navbar-toggler align-self-center"
            type="button"
            onClick={toggleSidebar}
          >
            <FontAwesomeIcon icon={faBars} />
          </button>
        )}

        <div
          className={cx(styles.extensionConsoleHeading, "d-none d-md-inline")}
        >
          Extension Console
        </div>

        <ul className="navbar-nav navbar-nav-right flex-grow-1 justify-content-end">
          {
            <Nav.Link className="px-3" target="_blank" href={serviceURL}>
              <FontAwesomeIcon icon={faExternalLinkAlt} className="mr-1" />
              Open Admin Console
            </Nav.Link>
          }
          {connected && email && <div className="text-black">{email}</div>}
        </ul>
      </div>
    </nav>
  );
};

export default Navbar;
