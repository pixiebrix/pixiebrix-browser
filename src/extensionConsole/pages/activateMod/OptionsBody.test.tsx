/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render, screen } from "@/extensionConsole/testHelpers";
import OptionsBody from "@/extensionConsole/pages/activateMod/OptionsBody";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";

beforeAll(() => {
  registerDefaultWidgets();
});

describe("Marketplace Activate Wizard OptionsBody", () => {
  it("renders a text field", async () => {});

  test("renders text field", async () => {
    const { asFragment } = render(
      <OptionsBody
        mod={{
          options: {
            schema: {
              properties: {
                textField: {
                  title: "Text Field",
                  type: "string",
                },
              },
            },
          },
        }}
      />,
      {
        initialValues: {
          optionsArgs: {},
        },
      },
    );

    await expect(screen.findByText("Text Field")).resolves.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
