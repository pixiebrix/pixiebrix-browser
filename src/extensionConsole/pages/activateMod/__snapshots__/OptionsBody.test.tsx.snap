// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marketplace Activate Wizard OptionsBody renders text field 1`] = `
<DocumentFragment>
  <div
    class="formGroup form-group"
  >
    <div
      class="mb-2 max-950 collapse"
    >
      <div
        class="annotationPlaceholder"
      />
    </div>
    <label
      class="label form-label"
      for="optionsArgs.textField"
    >
      Text Field
    </label>
    <div
      class="formField"
    >
      <div
        class="root"
      >
        <div
          class="field"
        >
          <input
            class="form-control"
            id="optionsArgs.textField"
            name="optionsArgs.textField"
            readonly=""
            value=""
          />
        </div>
        <div
          class="dropdown dropdown"
          data-test-selected="Exclude"
          data-testid="toggle-optionsArgs.textField"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle btn btn-link"
            type="button"
          >
            <span
              class="symbol"
            >
              <test-file-stub
                classname="root"
              />
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
